clippy_lints/src/bool_assert_comparison.rs

```diff
@@ -74,74 +77,115 @@ impl<'tcx> LateLintPass<'tcx> for BoolAssertComparison {
-                    suggestions.push((non_lit_expr.span, (!sugg).to_string()));
-                }
-
-                diag.multipart_suggestion(
-                    format!("replace it with `{non_eq_mac}!(..)`"),
-                    suggestions,
-                    Applicability::MachineApplicable,
-                );
-            },
-        );
+    if !is_impl_not_trait_with_bool_out(cx, non_lit_ty) {
+        // At this point the expression which is not a boolean
+        // literal does not implement Not trait with a bool output,
+        // so we cannot suggest to rewrite our code
+        return;
+    }
+
+    if !is_copy(cx, non_lit_ty) {
+        // Only lint with types that are `Copy` because `assert!(x)` takes
+        // ownership of `x` whereas `assert_eq(x, true)` does not
+        return;
+    }
+
+    let non_eq_mac = &macro_name[..macro_name.len() - 3];
+    span_lint_and_then(
+        cx,
+        BOOL_ASSERT_COMPARISON,
+        macro_call.span,
+        format!("used `{macro_name}!` with a literal bool"),
+        |diag| {
+            // assert_eq!(...)
+            // ^^^^^^^^^
+            let name_span = cx.sess().source_map().span_until_char(macro_call.span, '!');
+
+            let mut suggestions = vec![(name_span, non_eq_mac.to_string()), (lit_span, String::new())];
+
+            if bool_value ^ eq_macro {
+                let Some(sugg) = Sugg::hir_opt(cx, non_lit_expr) else {
+                    return;
+                };
+                suggestions.push((non_lit_expr.span, (!sugg).to_string()));
+            }
+
+            diag.multipart_suggestion(
+                format!("replace it with `{non_eq_mac}!(..)`"),
+                suggestions,
+                Applicability::MachineApplicable,
+            );
+        },
+    );
+}
+
+/// Checks for `assert!(a == b)` and `assert!(a != b)`
+fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, macro_call: &MacroCall, macro_name: &str) {
+    let Some((cond, _)) = find_assert_args(cx, expr, macro_call.expn) else {
+        return;
+    };
+    let ExprKind::Binary(Spanned { node, span }, lhs, rhs) = cond.kind else {
+        return;
+    };
+    // TODO: is `cond.span.from_expansion()` correct / needed?
+    if (node != BinOpKind::Eq && node != BinOpKind::Ne) || cond.span.from_expansion() {
+        return;
     }
+
+    let new_name = format!("{macro_name}_{}", if node == BinOpKind::Eq { "eq" } else { "ne" });
+    let msg = format!("replace it with `{new_name}!(..)`");
+    span_lint_and_then(
+        cx,
+        BOOL_ASSERT_COMPARISON,
+        macro_call.span,
+        format!("used `{macro_name}!` with an equality comparison"),
+        |diag| {
+            let macro_name_span = cx.sess().source_map().span_until_char(macro_call.span, '!');
+            // TODO: should this be `cond.span`, expanded to include preceding whitespace? If so, how?
+            let equality_span = Span::new(lhs.span.hi(), rhs.span.lo(), span.ctxt(), span.parent());
```

Comment:
We know that the `==` actually comes from the source code, so I would probably use `cond.span.with_lo(lhs.span.hi()).with_high(rhs.span.lo())`. But this should result in the same span, so either way should be fine :)